text/0112-server-runtime.md

```diff
@@ -0,0 +1,331 @@
+- `renderToString(element: ServerHTMLElement): string`: This method creates an LWC component tree synchronously and serialize it to string. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.
+
+This package injects mock DOM APIs in the `@lwc/engine-core` rendering engine. Those DOM APIs produces a lightweight DOM structure that can be serialized into a string by the `renderToString` method. As described in the Appendix, this package is also in charge of attaching on the global object a mock `CustomEvent`.
+
+## Drawbacks
+
+- Requires a complete rewrite of the current `@lwc/engine`
+- Might require substantial changes to the existing tools (jest preset, rollup plugins, ...) to load the right engine depending on the environment.
+
+## Alternatives
+
+### Using a DOM implementation in JavaScript before evaluating the engine
+
+Prior the evaluation of the LWC engine, we would evaluate a DOM implementation write in JavaScript (like [jsdom](https://github.com/jsdom/jsdom) or [domino](https://github.com/fgnass/domino)). The compelling aspect of this approach is that it requires almost no change to the engine to work.
+
+There are multiple drawbacks with this approach:
+
+- Performance: The engine only relies on a limited set of well-known APIs, leveraging a full DOM implementation for SSR would greatly reduce the SSR performance.
+- Predictability: By attaching the DOM interfaces on the global object, those APIs are not only exposed to the engine but also to the component author. Exposing such APIs to the component author might bring unexpected behavior when component code runs on the server.
+
+## Adoption strategy
+
+TBD
+
+## How we teach this
+
+TBD
+
+## Unresolved questions
+
+**How to load the right version of the engine when resolving the `lwc` identifier?**
+
+For example, when running in Node.js `lwc` might resolve to `@lwc/engine-server` for SSR purposes, but also to `@lwc/engine-dom` when running test via jest and jsdom.
+
+One way to solve this would be to import the platform-specific APIs from the appropriate module: `createElement` from `@lwc/engine-dom`, `createElement` and `renderToString` from `@lwc/engine-server` and use. Platform-agnostic APIs should be imported from `@lwc` (which becomes an alias to `@lwc/engine-core`).
+
+---
+
+## Appendix
+
+### DOM APIs usage
+
+We can break-down the current LWC DOM usages into 3 different categories:
+
+- [DOM constructors used by the engine](#dom-constructors-used-by-the-engine)
+- [DOM methods and accessors used by the engine](#dom-methods-and-accessors-used-by-the-engine)
+- [DOM constructors used by component authors](#dom-constructors-used-by-component-authors)
```

Comment:
component authors might use regular APIs as well, not just constructors

Reply:
What APIs are you referring to with "regular APIs"? I don't see any other APIs except the one listed in this section that might be involved in SSR.

Reply:
this item is about users code, in their constructor, and in their hook callbacks they can do anything. I believe this bullet point was about that... maybe I'm wrong...what is this item about?

Reply:
IMO, the only predictable way to make component render on a non-DOM environment is to restrict all the DOM API access when running in such environment. This would force component authors to write more defensive code if they want their components to be eligible for SSR. The same constraints are also enforced by the rest of the popular framework.

For example, if a component needs to attach an event handler on the `window` object in the `connectedCallback` the code needs to check first if the component is running in a browser.

```js
import { LightningElement } from 'lwc';

export default class App extends LightningElement {
  connectedCallback() {
    if (/* check if running in a browser */) {
      window.addEventListener('resize', () => { /* do something */ })
    }
  }
}
```

It's important to note that only a subset of the components written today will be capable to render on the server.

I also commented out the code checking if the component is running in the browser because I don't LWC should provide such mechanism. I would rather delegate this responsibility to the application framework.

Reply:
right, no controversy there... just I don't understand what "DOM constructors used by component authors" means?

Reply:
Those are DOM interface constructors that developers might need access even in a non-DOM enabled context. So far, I have found 2 constructors that need to be present on the global object: `CustomEvent` and `Event`. 

I am happy to change the title name for something more explicit.

Reply:
Ok, Ok, the important word was "interface", I understand now what you're saying. We are good.